### PR TITLE
[PyTorch] Add support for aten::quantize_linear and aten::dequantize

### PR DESCRIPTION
**Pull request by zrphercule**
This pr create basic support for aten::quantize_linear and aten::dequantize.
Todo: add unit test, the correctness of this pr is not tested yet.
Feel free to give some high level comments.

### CODE REVIEW

**Comment by nadavrot** [comment]
Cool @zrphercule .

File: torch_glow/tests/nodes/quantize_dequantize_test.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+from tests.utils import jitVsGlow
+
+
+def test_quantization_basic():
+    """Basic test of the PyTorch quantize and dequantize Node on Glow."""
+
+    def test_f(a, b):
+        q = torch.nn.quantized.Quantize(255, 0, torch.quint8)
+        dq = torch.nn.quantized.DeQuantize()
+        return dq(q(a.add(b)))
```
Comment (jackm321) [inline torch_glow/tests/nodes/quantize_dequantize_test.py]: Why does this test add then do quant-dequant on empty network? Should it be dq(q(a)) or put the add in the middle of the quantized section?

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 2]
```diff
@@ -1092,6 +1115,60 @@ llvm::Error PyTorchModelLoader::loadBatchNorm(const torch::jit::Node *ptNode) {
...
+                             getGlowNodeValue(inputs[QuantizeInputs::input]));
+
+  // scale
+  auto scaleHandle = glow::Handle<float>::createInvalidHandle();
+  ASSIGN_VALUE_OR_RETURN_ERR(
+      scaleHandle, getGlowConstantHandle<float>(inputs[QuantizeInputs::scale]));
+  float outScale = scaleHandle.raw(0);
+
+  // offset
+  auto offsetHandle = glow::Handle<int8_t>::createInvalidHandle();
```
Comment (rdzhabarov) [inline torch_glow/src/PyTorchModelLoader.cpp]: just checking on this. is offset int8? (typically it's int32)

Comment (zrphercule) [reply torch_glow/src/PyTorchModelLoader.cpp]: yeah it is int32...

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 3]
```diff
@@ -1167,6 +1190,56 @@ llvm::Error PyTorchModelLoader::loadBatchNorm(const torch::jit::Node *ptNode) {
...
+  ASSIGN_VALUE_OR_RETURN_ERR(
+	  outOffset, iValToInt(getGlowIValueForValue(inputs[QuantizeInputs::offset])));
+
+  // dtype, we only support quantize to int8 for now
+  int32_t outDtype;
+  ASSIGN_VALUE_OR_RETURN_ERR(outDtype, iValToInt(getGlowIValueForValue(inputs[QuantizeInputs::dtype])));
+
+  // Right now pytorch only has quint8 quantization support, therefore we only
+  // support uint8 as well.
+  RETURN_ERR_IF_NOT(outDtype == 13,
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: Can we use ScalarType::QUInt8 instead of 13 here?

**Comment by zrphercule** [comment]
> I'm curious, unlike c2, we should be able to easily pass quantized tensors in and out of glow fusion group right? Like quantize and/or dequantize can happen in pytorch jit and just quantized ops will happen in glow. Is that correct?

I think you are right, the output of quantize & input of dequantize are both quantized tensors in pytorch, which means so long as we have socket for quantized tensor, it should work. Also quantize/dequantize in pytorch may help us evade some problem (but still we need to do requantize everytime we use some ops in glow right?).
While I also think we should keep these two ops working and leave the choice to users. Since sometimes quantize is not the start of the model and dequantize is not the end of the model, we dont want the graph to be separated by these two ops.

**Comment by yinghai** [comment]
> I'm curious, unlike c2, we should be able to easily pass quantized tensors in and out of glow fusion group right? Like quantize and/or dequantize can happen in pytorch jit and just quantized ops will happen in glow. Is that correct?

The tricky part is how we are going to create a quantized pt tensor from glow quantized tensor and carry the quantization info. 
https://github.com/pytorch/glow/blob/aa4f6fe5ff380a0002567a8f05f8d0f76fc500ff/torch_glow/src/PyTorchCommon.cpp#L174-L175

File: torch_glow/tests/nodes/quantize_dequantize_test.py (hunk 1 above)
Comment (zrphercule) [reply torch_glow/tests/nodes/quantize_dequantize_test.py]: That would need a translation of quantized add operator as well, which introduced more non-stable stuff. I prefer to let the test of quantize-dequantize be like this, at least for now.

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 4]
```diff
@@ -1163,6 +1186,56 @@ Error PyTorchModelLoader::loadBatchNorm(const torch::jit::Node *ptNode) {
...
+                                            inputs[QuantizeInputs::offset])));
+
+  // dtype, we only support quantize to int8 for now
+  int32_t outDtype;
+  ASSIGN_VALUE_OR_RETURN_ERR(outDtype, iValToInt(getGlowIValueForValue(
+                                           inputs[QuantizeInputs::dtype])));
+
+  // Right now pytorch only has quint8 quantization support, therefore we only
+  // support uint8 as well.
+  RETURN_ERR_IF_NOT(outDtype == 13, "we only support to be quantized as uint8");
```
Comment (zrphercule) [inline torch_glow/src/PyTorchModelLoader.cpp]: oh this equal to 13 stuff I need to change.

File: torch_glow/src/PyTorchModelLoader.h [hunk 5]
```diff
@@ -302,6 +302,14 @@ class PyTorchModelLoader {
   /// \returns error on failure.
   Error loadBatchNorm(const torch::jit::Node *ptNode);
 
+  /// Load a PyTorch quantize_linear node.
+  /// \returns error on failure.
+  Error loadQuantize(const torch::jit::Node *ptNode);
+
+  /// Load a PyTorch dequantize_linear node.
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.h]: dequantize not dequantize_linear

File: torch_glow/tests/nodes/quantize_dequantize_test.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+def test_quantization_basic():
+    """Basic test of the PyTorch quantize and dequantize Node on Glow."""
+
+    # This only test if operator has been successfully projected
+    # Calculating process in glow will be optimized
+    def test_simple(a, b):
+        q = torch.nn.quantized.Quantize(1, 0, torch.quint8)
+        dq = torch.nn.quantized.DeQuantize()
+        return dq(q(a.add(b)))
```
Comment (jackm321) [inline torch_glow/tests/nodes/quantize_dequantize_test.py]: add a todo to move the add into the lowered graph?

Comment (zrphercule) [reply torch_glow/tests/nodes/quantize_dequantize_test.py]: I have added some comments before this test_simple already, but well add more wont bite ;)

**Pull request closed without merging**